imap_processing/spice/pointing_frame.py

```diff
@@ -100,72 +183,71 @@ def create_pointing_frame(
+        # There were rounding errors when using spiceypy.pxform
+        # so np.ceil and np.floor were used to ensure the start
+        # and end times were within the ck range.
+        et_times = np.linspace(
+            np.ceil(pointing_start_et * 1e6) / 1e6,
+            np.floor(pointing_end_et * 1e6) / 1e6,
+            int(num_samples),
+        )
+
+        # Get the average quaternions for the pointing
+        q_avg = _average_quaternions(et_times)
+
+        # Create a rotation matrix
+        rotation_matrix = _create_rotation_matrix(q_avg)
+
+        # Convert the rotation matrix to a quaternion.
+        # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.m2q
+        quaternion[i_pointing] = spiceypy.m2q(rotation_matrix)
 
-            # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.sce2c
-            # Convert start and end times to SCLK.
-            sclk_begtim = spiceypy.sce2c(int(id_imap_sclk), et_times[0])
-            sclk_endtim = spiceypy.sce2c(int(id_imap_sclk), et_times[-1])
+        # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.sce2c
+        # Convert start and end times to SCLK ticks.
+        sclk_start[i_pointing] = spiceypy.sce2c(int(id_imap_sclk), pointing_start_et)
+        sclk_end[i_pointing] = spiceypy.sce2c(int(id_imap_sclk), pointing_end_et)
 
-            # Create the pointing frame kernel.
-            # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.ckw02
-            spiceypy.ckw02(
-                # Handle of an open CK file.
-                handle,
-                # Start time of the segment.
-                sclk_begtim,
-                # End time of the segment.
-                sclk_endtim,
-                # Pointing frame ID.
-                int(id_imap_dps),
-                # Reference frame.
-                "ECLIPJ2000",  # Reference frame
-                # Identifier.
-                "IMAP_DPS",
-                # Number of pointing intervals.
-                1,
-                # Start times of individual pointing records within segment.
-                # Since there is only a single record this is equal to sclk_begtim.
-                np.array([sclk_begtim]),
-                # End times of individual pointing records within segment.
-                # Since there is only a single record this is equal to sclk_endtim.
-                np.array([sclk_endtim]),  # Single stop time
-                # Average quaternion.
-                q_avg,
-                # 0.0 Angular rotation terms.
-                np.array([0.0, 0.0, 0.0]),
-                # Rates (seconds per tick) at which the quaternion and
-                # angular velocity change.
-                np.array([1.0]),
-            )
+    return np.array(
+        list(zip(sclk_start, sclk_end, quaternion, pointing_id)),
+        dtype=POINTING_SEGMENT_DTYPE,
+    )
```

Comment:
If you are creating this structured array here already could you create this above initially like I mentioned and then fill in the pieces as you go?

```python
pointing_segments = np.zeros(n_pointings, dtype=POINTING_SEGMENT_DTYPE)

pointing_segments["sclk_start"][i_pointing] = spiceypy.sce2c(int(id_imap_sclk), pointing_start_et)

return pointing_segments
```